### PR TITLE
Add new bidder: Impactify

### PR DESCRIPTION
**Pull request by yevhenii-viktorov**
New bidder contains as follows:
- https://github.com/prebid/prebid-server/pull/2004
- https://github.com/prebid/prebid-server/pull/2053

### CODE REVIEW

File: src/main/resources/static/bidder-params/impactify.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      "type": "string",
+      "description": "The appId of your website"
+    },
+    "format": {
+      "type": "string",
+      "description": "The format of the ad"
+    },
+    "style": {
+      "type": "string",
+      "description": "The style of the ad"
```
Comment (And1sS) [inline src/main/resources/static/bidder-params/impactify.json:17]: In go they also have minLength restrictions. Please, update this schema to be in line with go.

File: src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+                .willReturn(BigDecimal.TEN);
+
+        final BidRequest bidRequest = givenBidRequest(
+                impCustomizer -> impCustomizer.bidfloor(BigDecimal.ONE).bidfloorcur("EUR"));
+
+        //when
+        Result<List<HttpRequest<BidRequest>>> result = impactifyBidder.makeHttpRequests(bidRequest);
+
+        //then
+        assertThat(result.getErrors()).hasSize(0);
```
Comment (And1sS) [inline src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java:90]: Please, replace with call to i`sEmpty()`, also, please, check for similar occurences.

File: src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    @Test
+    public void makeHttpRequestsShouldConvertCurrencyIfNotDefault() {
+        // given
+        given(currencyConversionService.convertCurrency(any(), any(), anyString(), anyString()))
+                .willReturn(BigDecimal.TEN);
+
+        final BidRequest bidRequest = givenBidRequest(
+                impCustomizer -> impCustomizer.bidfloor(BigDecimal.ONE).bidfloorcur("EUR"));
+
+        //when
```
Comment (And1sS) [inline src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java:86]: Very minor, plese, add space between // and [given|when|then]. Also, pls, check for similar occurences.

File: src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+    @Test
+    public void makeHttpRequestsShouldReturnValidBidResponseWithAllHeadersExceptIpv6() {
+        //given
+        final BidRequest bidRequest = givenBidRequest(
+                bidRequestCustomizer -> bidRequestCustomizer
+                        .device(givenDevice(deviceCustomizer -> deviceCustomizer.ua("ua").ip("ip").ipv6("ipv6")))
+                        .site(Site.builder().page("https://proper.web.site").build())
+                        .user(User.builder().buyeruid("buyer_user_uid").build()),
+                singletonList(impCustomizer -> impCustomizer
```
Comment (And1sS) [inline src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java:125]: Time to replace old facioned calls to singletonList with shiny new List.of()? ;)

File: src/test/java/org/prebid/server/it/ImpactifyTest.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import static com.github.tomakehurst.wiremock.client.WireMock.post;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
+import static java.util.Collections.singletonList;
+
+@RunWith(SpringRunner.class)
+public class ImpactifyTest extends IntegrationTest {
+
+    @Test
+    public void openrtb2AuctionShouldRespondWithBidsFromImpactify() throws IOException, JSONException {
+        //given
```
Comment (And1sS) [inline src/test/java/org/prebid/server/it/ImpactifyTest.java:23]: // given

**Review by And1sS — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        return Result.withValue(HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(resolveEndpoint())
+                .headers(constructHeaders(updatedBidRequest))
+                .body(mapper.encode(updatedBidRequest))
+                .payload(updatedBidRequest)
+                .build());
+    }
+
+    private static boolean isBidFloorValid(BigDecimal bidFloor, String bidFloorCur) {
```
[the remark this answers is not in the document]
Comment (yevhenii-viktorov) [reply src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:105]: In context of code that goes after it, should be good

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        final String deviceIpv6 = device != null ? device.getIpv6() : null;
+        final Site site = bidRequest.getSite();
+        final String sitePage = site != null ? site.getPage() : null;
+        final User user = bidRequest.getUser();
+        final String userUid = user != null ? user.getBuyeruid() : null;
+        final MultiMap headers = HttpUtil.headers();
+
+        headers.set(HttpUtil.X_OPENRTB_VERSION_HEADER, X_OPENRTB_VERSION);
+        headers.set(HttpUtil.CONTENT_TYPE_HEADER, HttpUtil.APPLICATION_JSON_CONTENT_TYPE);
+        headers.set(HttpUtil.ACCEPT_HEADER, HttpHeaderValues.APPLICATION_JSON);
```
[the remark this answers is not in the document]
Comment (yevhenii-viktorov) [reply src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:141]: Yep, found out about it yesterday, forgot to change this.

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    private static MultiMap constructHeaders(BidRequest bidRequest) {
+        final Device device = bidRequest.getDevice();
+        final String deviceUa = device != null ? device.getUa() : null;
+        final String deviceIpv4 = device != null ? device.getIp() : null;
+        final String deviceIpv6 = device != null ? device.getIpv6() : null;
+        final Site site = bidRequest.getSite();
+        final String sitePage = site != null ? site.getPage() : null;
+        final User user = bidRequest.getUser();
+        final String userUid = user != null ? user.getBuyeruid() : null;
+        final MultiMap headers = HttpUtil.headers();
```
[the remark this answers is not in the document]
Comment (yevhenii-viktorov) [reply src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:137]: I'd say then we can inline some of these if not all of them.

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+            final ExtImpImpactify extImpImpactify;
+            try {
+                extImpImpactify = mapper.mapper()
+                        .convertValue(imp.getExt(), IMPACTIFY_EXT_TYPE_REFERENCE)
+                        .getBidder();
+            } catch (IllegalArgumentException e) {
+                return Result.withError(
+                        BidderError.badInput("Unable to decode the impression ext for id: " + imp.getId()));
+            }
```
[the remark this answers is not in the document]
Comment (yevhenii-viktorov) [reply src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:82]: You have missed if statement, but I get an idea :)

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+                if (imp.getBanner() != null) {
+                    return BidType.banner;
+                }
+                if (imp.getVideo() != null) {
+                    return BidType.video;
+                }
+                throw new PreBidException(String.format("Unknown impression type for ID: '%s'", impId));
+            }
+        }
+        throw new PreBidException(String.format("Failed to find impression for ID: '%s'", impId));
```
[the remark this answers is not in the document]
Comment (yevhenii-viktorov) [reply src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:213]: Yea, but they do not factor case where imp not found, but I agree that we should follow existing functionality.

File: src/main/resources/bidder-config/impactify.yaml [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    endpoint: https://sonic.impactify.media/bidder
+    meta-info:
+      maintainer-email: support@impactify.io
+      site-media-types:
+        - banner
+        - video
+      supported-vendors:
+      vendor-id: 606
+    usersync:
+      url: https://sonic.impactify.media/static/cookie_sync.html?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}&redirect_url={{.RedirectURL}}
```
[the remark this answers is not in the document]
Comment (yevhenii-viktorov) [reply src/main/resources/bidder-config/impactify.yaml:12]: They have same thing in go repo

File: src/main/resources/bidder-config/impactify.yaml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+adapters:
+  impactify:
+    endpoint: https://sonic.impactify.media/bidder
+    meta-info:
+      maintainer-email: support@impactify.io
```
Comment (snahornyi) [inline src/main/resources/bidder-config/impactify.yaml:5]: According to project style pls add 
```yaml
app-media-types:
```

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+                .body(mapper.encodeToBytes(updatedBidRequest))
+                .payload(updatedBidRequest)
+                .build());
+    }
+
+    private static boolean shouldConvertBidFloor(BigDecimal bidFloor, String bidFloorCur) {
+        return bidFloor != null
+                && StringUtils.isNotEmpty(bidFloorCur)
+                && BigDecimal.ZERO.compareTo(bidFloor) <= 0
+                && !bidFloorCur.equalsIgnoreCase(BIDDER_CURRENCY);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:91]: They check only for price more then zero. Pls update this function
```java
BidderUtil.isValidPrice(bidFloor) && !StringUtils.equalsIgnoreCase(bidFloorCur, BIDDER_CURRENCY);
```

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+            BigDecimal bidFloor = imp.getBidfloor();
+            try {
+                if (shouldConvertBidFloor(bidFloor, imp.getBidfloorcur())) {
+                    bidFloor = convertBidFloorCurrency(bidFloor, request, imp.getId(), imp.getBidfloorcur());
+                }
+                updatedImps.add(updateImp(imp, bidFloor));
+            } catch (PreBidException e) {
+                return Result.withError(BidderError.badInput(e.getMessage()));
+            }
+        }
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:71]: What if
```java
for (Imp imp : request.getImp()) {
            try {
                final BigDecimal resolvedBidFloor = resolveBidFloor(request, imp, imp.getBidfloor());
                updatedImps.add(updateImp(imp, resolvedBidFloor));
            } catch (PreBidException e) {
                return Result.withError(BidderError.badInput(e.getMessage()));
            }
        }
        
....

private BigDecimal resolveBidFloor(BidRequest request, Imp imp, BigDecimal bidFloor) {
        return shouldConvertBidFloor(bidFloor, imp.getBidfloorcur())
                ? convertBidFloorCurrency(bidFloor, request, imp.getId(), imp.getBidfloorcur())
                : imp.getBidfloor();
    }
```

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    private static BidderBid resolveBidderBid(Bid bid, String currency, List<Imp> imps, List<BidderError> errors) {
+        final BidType bidType;
+        try {
+            bidType = getBidType(bid.getImpid(), imps);
+        } catch (PreBidException e) {
+            errors.add(BidderError.badServerResponse(e.getMessage()));
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:183]: Actually from go, they fail with first error

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+                if (shouldConvertBidFloor(bidFloor, imp.getBidfloorcur())) {
+                    bidFloor = convertBidFloorCurrency(bidFloor, request, imp.getId(), imp.getBidfloorcur());
+                }
+                updatedImps.add(updateImp(imp, bidFloor));
+            } catch (PreBidException e) {
+                return Result.withError(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        final BidRequest updatedBidRequest = request.toBuilder()
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:73]: As you hava `updateImp` seems something similar should be for bidRequest

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    private static MultiMap constructHeaders(BidRequest bidRequest) {
+        final MultiMap headers = HttpUtil.headers();
+
+        headers.set(HttpUtil.X_OPENRTB_VERSION_HEADER, X_OPENRTB_VERSION);
+
+        final Device device = bidRequest.getDevice();
+        if (Objects.nonNull(device)) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:130]: According to project style pls write `device != null`

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.USER_AGENT_HEADER, device.getUa());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, device.getIp());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, device.getIpv6());
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:133]: From go `Add IPv4 or IPv6 if available`

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+        }
+
+        final Site site = bidRequest.getSite();
+        final String sitePage = site != null ? site.getPage() : null;
+        HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.REFERER_HEADER, sitePage);
+
+        final User user = bidRequest.getUser();
+        final String userUid = user != null ? user.getBuyeruid() : null;
+        if (Objects.nonNull(user) && StringUtils.isNotEmpty(userUid)) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:142]: Seems just `StringUtils.isNotEmpty(userUid)` would be enough

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+        final Device device = bidRequest.getDevice();
+        if (Objects.nonNull(device)) {
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.USER_AGENT_HEADER, device.getUa());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, device.getIp());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, device.getIpv6());
+        }
+
+        final Site site = bidRequest.getSite();
+        final String sitePage = site != null ? site.getPage() : null;
+        HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.REFERER_HEADER, sitePage);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:138]:
```java
final Site site = bidRequest.getSite();
        if (site != null) {
            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.REFERER_HEADER, site.getPage());
        }
```

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+                    "Unable to convert provided bid floor currency from %s to %s for imp `%s`",
+                    bidFloorCur, BIDDER_CURRENCY, impId));
+        }
+    }
+
+    private Imp updateImp(Imp imp, BigDecimal bidFloor) {
+        return imp.toBuilder()
+                .bidfloorcur(BIDDER_CURRENCY)
+                .bidfloor(bidFloor)
+                .ext(mapper.mapper().valueToTree(ExtPrebid.of(null, parseExtImp(imp))))
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:110]: That's not what was done in go, check again ext processing, you can found expected result in go `sample_banner.json`

File: src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java [hunk 22]
```diff
@@ -0,0 +1,326 @@
...
+
+        // then
+        assertThat(result.getErrors()).hasSize(1)
+                .extracting(BidderError::getMessage)
+                .containsExactly("Unable to convert provided bid floor currency from EUR to USD for imp `123`");
+
+    }
+
+    @Test
+    public void makeHttpRequestsShouldReturnValidBidResponseWithAllHeadersExceptIpv6() {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java:116]: From test name exceptIpv6, in expected result `tuple(HttpUtil.X_FORWARDED_FOR_HEADER.toString(), "ipv6")`. Seems something wrong

File: src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java [hunk 23]
```diff
@@ -0,0 +1,326 @@
...
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getBidfloor, Imp::getBidfloorcur)
+                .containsExactly(tuple(BigDecimal.TEN, "USD"));
+    }
+
+    @Test
+    public void makeHttpRequestsWithValidDataWillThrowExceptionOnCurrencyConversion() {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java:97]: Not `will` but `should`. Also pls re-read test name, it's too weird. Pls change it. Pls check similar occurrences

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java (hunk 14 above)
Comment (yevhenii-viktorov) [reply src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:71]: Yea, that is better, obviously! Thank you!

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java (hunk 15 above)
Comment (yevhenii-viktorov) [reply src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:183]: That is why I did this:
'try {
bidType = getBidType(bid.getImpid(), imps);
} catch (PreBidException e) {
errors.add(BidderError.badServerResponse(e.getMessage()));
return null;
}'

**Review by And1sS — CHANGES_REQUESTED** [review]

**Review by And1sS — APPROVED** [review]

**Review by rpanchyk — DISMISSED** [review]

File: src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java [hunk 24]
```diff
@@ -0,0 +1,386 @@
...
+    }
+
+    private static HttpCall<BidRequest> givenHttpCall(BidRequest bidRequest, String body) {
+        return HttpCall.success(
+                HttpRequest.<BidRequest>builder().payload(bidRequest).build(),
+                HttpResponse.of(200, null, body),
+                null);
+    }
+
+    private static ObjectNode createObjectNode() {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java:377]: Pls use better name for this method. Also pls use ext java model

File: src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java [hunk 25]
```diff
@@ -0,0 +1,386 @@
...
+        final Result<List<BidderBid>> result = impactifyBidder.makeBids(httpCall, bidRequest);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue())
+                .extracting(BidderBid::getBid)
+                .containsOnly(Bid.builder()
+                        .impid("123")
+                        .build());
+
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java:310]: Useless /n

File: src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java [hunk 26]
```diff
@@ -0,0 +1,386 @@
...
+    private static BidRequest givenBidRequest(UnaryOperator<Imp.ImpBuilder> impCustomizer) {
+        return givenBidRequest(identity(), impCustomizer);
+    }
+
+    private static Device givenDevice(UnaryOperator<Device.DeviceBuilder> deviceCustomizer) {
+        return deviceCustomizer.apply(Device.builder()).build();
+    }
+
+    private static Imp givenImp(Function<Imp.ImpBuilder, Imp.ImpBuilder> impCustomizer) {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/impactify/ImpactifyBidderTest.java:354]: UnaryOperator

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java [hunk 27]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    private ObjectNode createImpExtObjectNode(ExtImpImpactify impExt) {
+        final ObjectNode modifiedImpExtBidder = mapper.mapper().createObjectNode();
+
+        if (impExt != null) {
+            modifiedImpExtBidder.set("appId", TextNode.valueOf(impExt.getAppId()));
+            modifiedImpExtBidder.set("format", TextNode.valueOf(impExt.getFormat()));
+            modifiedImpExtBidder.set("style", TextNode.valueOf(impExt.getStyle()));
+        }
```
[the remark this answers is not in the document]
Comment (yevhenii-viktorov) [reply src/main/java/org/prebid/server/bidder/impactify/ImpactifyBidder.java:134]: The implemented this part in other way

**Review by snahornyi — APPROVED** [review]

**Pull request merged**